internal/services/v1/permissions.go

```diff
@@ -535,86 +536,157 @@ func (ps *permissionServer) LookupSubjects(req *v1.LookupSubjectsRequest, resp v
+	}
+
+	internalResponseCursor := &impl.DecodedCursor{
+		VersionOneof: &impl.DecodedCursor_V1{
+			V1: &impl.V1Cursor{
+				Revision:              atRevision.String(),
+				CallAndParametersHash: lsRequestHash,
+			},
+		},
+	}
+
+	for {
+		countSubjectsFound := 0
 
-		for _, foundSubject := range foundSubjects.FoundSubjects {
-			excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream := dispatchpkg.NewHandlingDispatchStream(ctxWithCancel, func(result *dispatch.DispatchLookupSubjectsResponse) error {
+			foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
+			if !ok {
+				return fmt.Errorf("missing resource ID in returned LS")
 			}
 
-			excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+			for _, foundSubject := range foundSubjects.FoundSubjects {
+				// Skip wildcards if requested they be skipped.
+				if req.WildcardOption == v1.LookupSubjectsRequest_WILDCARD_OPTION_EXCLUDE_WILDCARDS && foundSubject.SubjectId == tuple.PublicWildcard {
+					continue
+				}
+
+				excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
+				excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
+				for _, excludedSubject := range foundSubject.ExcludedSubjects {
+					excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+
+					resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+					if err != nil {
+						return fmt.Errorf("error when resolving excluded subject: %w", err)
+					}
+
+					if resolvedExcludedSubject == nil {
+						continue
+					}
+
+					excludedSubjects = append(excludedSubjects, resolvedExcludedSubject)
+				}
+
+				subject, err := foundSubjectToResolvedSubject(ctx, foundSubject, caveatContext, ds)
+				if err != nil {
+					return fmt.Errorf("error when resolving subject: %w", err)
+				}
+				if subject == nil {
+					continue
+				}
+
+				// NOTE: we need to recompute the cursor here because we get multiple results back from DispatchLookupSubjects
+				// in one message.
+				dispatchCursor, err := graph.CursorForFoundSubjectID(subject.SubjectObjectId, result.AfterResponseCursor)
 				if err != nil {
 					return err
 				}
 
-				if resolvedExcludedSubject == nil {
-					continue
+				// Update the existing internal cursor for encoding.
+				internalResponseCursor.GetV1().DispatchVersion = dispatchCursor.DispatchVersion
+				internalResponseCursor.GetV1().Sections = dispatchCursor.Sections
```

Comment:
Is the stream handler executed serially over N results? Thinking if there is a data race here, the variable is in scope of the LR controller

Reply:
Yes, its serial